Tablets 4.x

Bouncheck opened the pull request:
(Copied over from commit message)
Introduces basic tablets support for version 4.x of the driver.
Metadata about tablets will be kept in TabletMap that gets continuously updated
through the tablets-routing-v1 extension. Each time the BoundStatement targets 
the wrong node and shard combination the server supporting tablets should
respond with tablet metadata inside custom payload of its response. 
This metadata will be transparently processed and used for future queries.

Tablets metadata will by enabled by default. Until now driver was using
TokenMaps to choose replicas and appropriate shards. Having a token was enough
information to do that. Now driver will first attempt tablet-based lookup
and only after failing to find corresponding tablet it will defer to TokenMap
lookup. Since to find a correct tablet besides the token we need the keyspace
and table names, many of the methods were extended to also accept those
as parameters.

RequestHandlerTestHarness was adjusted to mock also MetadataManager.
Before it used to mock only `session.getMetadata()` call but the same can
be obtained by `context.getMetadataManager().getMetadata()`. Using the
second method was causing test failures.

Fixes #268

roydahan commented:
@Bouncheck can you please describe the current status of it?
Where we're in terms of code completion / testing / what's left / any estimation for completion?

Also, if review is already required or not and if any blockers exist at the moment.

Bouncheck commented:
Currently I've made almost complete switch to the solution that uses custom payloads instead of querying system.tablets.
I've ran some tests on local 4 node cluster and so far there were no issues with reading the payloads and selecting the correct replica node. The only thing current code lacks is choosing a shard correctly.
I think the code will be complete during this week.

roydahan commented:
@Bouncheck can you please update what’s the status and what’s this PR waiting for?

Bouncheck commented:
Currently it's waiting for review. I've done some manual testing, but haven't done any stress testing under heavy load.

roydahan commented:
Thanks @Bouncheck, can you please check the failing CI check?
Also, please ping someone from the team to help you with a peer review.

Regarding stress testing, I don't know if we have a tool/app that utilizes this driver version, IIRC there was a problem to build c-s with this driver version. However, if it's not the case anymore or we can utilize newer c-s from the upstream that we can build with this driver, we can have a longevity test using SCT.

Bouncheck commented:
Will do. At first i thought 2024.1.2 failures were a one off, since it looked liked there's a lot of similar test errors relating to starting the cluster (and only for that scylla version), but it seems they reproduce after all.

roydahan commented:
@Bouncheck what's the status of this PR?

@Lorak-mmk can you please help reviewing this?

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/PreparedStatement.java:
```diff
@@ -58,6 +59,10 @@ public interface PreparedStatement {
   @NonNull
   ColumnDefinitions getVariableDefinitions();
 
+  /** Table name inferred from {@link PreparedStatement#getVariableDefinitions()}. */
+  @Nullable
+  CqlIdentifier getTable();
+
```
Why `PreparedStatement` now has `getTable` method, but not `getKeyspace` method? It's inconsistent and makes the rest of the code more confusing.

Lorak-mmk replied:
I see now that `interface Request` has `getKeyspace` and `getRoutingKeyspace`. Why is table getter implemented here and keyspace getter there?

Bouncheck replied:
I've renamed this to `inferTable()` for more clarity. `getKeyspace` and `getRoutingKeyspace` were already implemented. I've implemented `inferTable()` here because I wanted to go as specific as possible, without modifying superclasses I don't need to.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/TabletMap.java:
```diff
@@ -0,0 +1,39 @@
+   */
+  public void addTablet(String keyspace, String table, TupleValue tupleValue);
+
+  /**
+   * Returns UUIDs of replicas holding tablets for given token. May contain stale data that
+   * eventually will be updated. To get actual Node instances from UUIDs see {@link
+   * Metadata#getNodes()}.
+   *
+   * @param keyspace tablet's keyspace
+   * @param table tablet's table
+   * @param token target token
+   * @return Set of {@link Node} that holds target tablets.
+   */
+  public Set<UUID> getReplicas(String keyspace, String table, long token);
```
I wonder, why `getReplicas` instead of `getTablet`? Former is more general.
Now if you wan't to get a `Tablet` you need to call `getMapping` and bascially reimplement lookup logic present in `getReplicas`.

Bouncheck replied:
Makes sense, I've modified `TabletMap` to expose `Tablet getTablet(...)` instead.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -284,6 +296,103 @@ private Token getRoutingToken(Statement statement) {
     return tokenMap == null ? null : ((DefaultTokenMap) tokenMap).getTokenFactory().hash(key);
   }
 
+  private CqlIdentifier getTabletRoutingKeyspace(Statement statement) {
+    if (statement == null) {
+      if (initialStatement == null) {
+        return null;
+      }
+      statement = initialStatement;
+    }
+    ColumnDefinitions cdefs = null;
+    CqlIdentifier result = null;
+    if (statement instanceof BoundStatement) {
+      cdefs = ((BoundStatement) statement).getPreparedStatement().getVariableDefinitions();
+    } else if (statement instanceof PreparedStatement) {
+      cdefs = ((PreparedStatement) statement).getVariableDefinitions();
+    }
+    if (cdefs != null && cdefs.size() > 0) {
+      result = cdefs.get(0).getKeyspace();
+    }
+    if (result == null) {
+      return keyspace;
+    } else {
+      return result;
+    }
+  }
```
I don't know Java Driver enough to understand this code and there are pretty much no comments in this whole class.
- How does keyspace differ from routing keyspace?
- What is `initialStatement`?
- Why do we need this method? `getReplicas` already has code to retrieve keyspace for statement

Bouncheck replied:
I've simplified this method since, this one was kinda bad. Hopefully the new one makes more sense.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -284,6 +296,103 @@ private Token getRoutingToken(Statement statement) {
+          statement);
+      return null;
+    }
+    if (tabletMap == null) {
```
Can this happen? Why can't we guarantee this to always be non-null?

Bouncheck replied:
Removed this. I think this is a remnant of times when it could be null. Right now metadata should always at least hold an empty map.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -284,6 +296,103 @@ private Token getRoutingToken(Statement statement) {
+          node);
+      return null;
+    }
+    if ((getTabletRoutingKeyspace(statement) == null && keyspace == null)
+        || getTabletRoutingTable(statement) == null) {
+      return null;
+    }
+    UUID targetHostUuid = node.getHostId();
+    long tokenValue = ((TokenLong64) token).getValue();
+    String statementKeyspace =
+        getTabletRoutingKeyspace(statement) != null
+            ? getTabletRoutingKeyspace(statement).asInternal()
+            : keyspace.asInternal();
```
Why do we need to check `keyspace` here manually? Can't `getTabletRoutingKeyspace` handle it?

Bouncheck replied:
In a newer version it now does handle it

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -284,6 +296,103 @@ private Token getRoutingToken(Statement statement) {
+          statementTable);
+      return null;
+    }
+    Tablet row = targetTablets.ceiling(DefaultTabletMap.DefaultTablet.malformedTablet(tokenValue));
+    if (row != null && row.getFirstToken() < tokenValue) {
```
Why not use `tabletMap::getReplicas`?

Bouncheck replied:
Fixed, using `TabletMap::getTablet` now

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java:
```diff
@@ -288,14 +292,21 @@ protected Set<Node> getReplicas(@Nullable Request request, @Nullable Session ses
     ByteBuffer key;
     Partitioner partitioner = null;
+
+    if (request instanceof BoundStatement) {
+      table = ((BoundStatement) request).getPreparedStatement().getTable();
+    }
```
Are we only able to do this on `BoundStatement`s? Why can `PreparedStatement` be handled here too?

Bouncheck replied:
I think right now yes. `PreparedStatement` by itself won't appear here because it does not subclass `Request`. I am not sure if there are other types of `Request`s i should be mindful of with Tablets.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java:
```diff
@@ -321,6 +332,31 @@ protected Set<Node> getReplicas(@Nullable Request request, @Nullable Session ses
+                  .getTokenFactoryRegistry()
+                  .tokenFactoryFor(maybeTokenMap.get().getPartitionerName()));
+      Token t = (token != null ? token : p.hash(key));
+      if (t instanceof TokenLong64) {
+        Map<UUID, Node> currentNodes = session.getMetadata().getNodes();
+        Set<UUID> replicaUUIDs =
+            tabletMap.getReplicas(
+                keyspace.asInternal(), table.asInternal(), ((TokenLong64) t).getValue());
+        Set<Node> replicas = new HashSet<>();
+        for (UUID replica : replicaUUIDs) {
+          if (currentNodes.containsKey(replica)) {
+            replicas.add(currentNodes.get(replica));
+          }
+        }
+        if (!replicas.isEmpty()) {
+          return replicas;
+        }
+      }
```
I think we should distinguish 2 cases here:
- Table is a Tablet table (so it is present in `tabletMap`) , but we don't have information for this Token. In this case imo we shouldn't fall back to `TokenMap` routing because it is totally irrelevant to a tablet table
- Table is vnode-based table: in this case yes, we should fall back to  `TokenMap` routing

Bouncheck replied:
Makes sense, but I've left this flow as it is for now. I'd like to fix that with later expanding `KeyspaceMetadata` and `TableMetadata` classes, but that is probably something more suitable for another PR.
Right now stupid fallbacks should happen only once for each tablet.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java:
```diff
@@ -288,14 +292,21 @@ protected Set<Node> getReplicas(@Nullable Request request, @Nullable Session ses
     if (!maybeTokenMap.isPresent()) {
       return Collections.emptySet();
     }
+    TabletMap tabletMap = context.getMetadataManager().getMetadata().getTabletMap();
```
When can it happen that `maybeTokenMap` is not present?
If it is happen we return always, even if table is tablet-based (in which case we have no need for `TokenMap`).
In Rust driver we split similar method into 2 branches of a big if - one for tablets and one for vnodes.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandler.java:
```diff
@@ -309,9 +418,20 @@ private void sendRequest(
     Node node = retriedNode;
     DriverChannel channel = null;
     if (node == null
-        || (channel = session.getChannel(node, logPrefix, getRoutingToken(statement))) == null) {
+        || (channel =
+                session.getChannel(
+                    node,
+                    logPrefix,
+                    getRoutingToken(statement),
+                    getShardFromTabletMap(statement, node, getRoutingToken(statement))))
+            == null) {
       while (!result.isDone() && (node = queryPlan.poll()) != null) {
-        channel = session.getChannel(node, logPrefix, getRoutingToken(statement));
+        channel =
+            session.getChannel(
+                node,
+                logPrefix,
+                getRoutingToken(statement),
+                getShardFromTabletMap(statement, node, getRoutingToken(statement)));
```
Do I understand this code correctly: first we lookup replicas for a tablet in load balancing and then we do this here again by calling `getShardFromTabletMap`, introducing another lookup and more failure possibilities?

Maybe a better option is to change queryPlan to optional store shards too?

Bouncheck replied:
Yes, we need to do 2 lookups. 
Extending queryPlans to include information about both node and shard is a good idea but it also is something I'd rather try in another PR. There are several classes to change, tests to adjust and probably some stuff I'm not seeing yet to adjust for backwards compatibility.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java:
```diff
@@ -321,6 +332,31 @@ protected Set<Node> getReplicas(@Nullable Request request, @Nullable Session ses
+              : context
+                  .getTokenFactoryRegistry()
+                  .tokenFactoryFor(maybeTokenMap.get().getPartitionerName()));
+      Token t = (token != null ? token : p.hash(key));
```
In case of fallback to vnode-based routing we'll discard results of `p.hash(key)` and unnecessarily calculate them again in `tokenMap.getReplicas(keyspace, partitioner, key)`

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/AddTabletRefresh.java:
I've read this file and I have no idea what it des and what is it used for. Maybe some comment would be a good idea fur future readers?

Bouncheck replied:
Added comment in a new version. In a new version it's just an implementation of MetadataRefresh that handles just inserting a Tablet into TabletMap.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,267 @@
+  public Set<UUID> getReplicas(String keyspace, String table, long token) {
+    KeyspaceTableNamePair key = new KeyspaceTableNamePair(keyspace, table);
+
+    if (mapping == null) {
+      logger.trace("This tablets map is null. Returning empty set.");
+      return Collections.emptySet();
+    }
```
Why it needs to be checked here? Can't you make sure that this field is never null?

Bouncheck replied:
Yeah, seems unnecessary. Removed this.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,267 @@
+      replicas.add(hostShardPair);
+    }
+
+    // Working on a copy to avoid concurrent modification of the same set
+    NavigableSet<Tablet> existingTablets =
+        new TreeSet<>(mapping.computeIfAbsent(ktPair, k -> new TreeSet<>()));
```
Wait wait wait wait..... we need to copy the whole tablet list for the table (which afaik may contain up to 100k tablets) for each tablet insertion? So inserting N tablet's has the complexity O(N^2)? With N being 10^5 this is 10^10 operations. This is not acceptable

Bouncheck replied:
10^5 indeed does sound like a lot. Removed the copying, since it seems unnecessary in current solution anyway.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,267 @@
+    existingTablets.add(new DefaultTablet(keyspace, null, table, firstToken, lastToken, replicas));
+
+    // Set the updated result in the main map
+    mapping.put(ktPair, existingTablets);
```
You said that working on a copy prevents concurrent modifications, but it looks like it introduces other serious problem. This whole method:
1. Clones mapping
2. Modifes mapping
3. Replaces mapping

What happends if 2 calls are made concurrently? Both will copy the field, calculate new mapping and then replace it - one of updates will be lost. Considering that tablet update is not a cheap operation and that there might be a lot of them after session initialization (making this race very likely) it doesn't seem like a good solution.

Bouncheck replied:
This version was bugged, but a revised one makes modifications sequentially. I've made it so MetadataManager handles those modifications like other metadata changes. This has some room for improvement but it's probably good enough for now.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    // Add new (now) non-overlapping tablet
+    existingTablets.add(new DefaultTablet(keyspace, null, table, firstToken, lastToken, replicas));
```
Can't you create this tablet earlier and use it during sweeps instead of `malformedTablet`s?
Then you could introduce `overlap` method on `Tablet` and use it in sweeps instead of comparing tokens directly, resulting in much more readable code.

Bouncheck replied:
Removed use of `malformedTablet` from here

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/pool/ChannelPool.java:
```diff
@@ -167,18 +167,23 @@ public DriverChannel next() {
-            ? singleThreaded.shardingInfo.shardId(routingKey)
-            : ThreadLocalRandom.current().nextInt(channels.length);
+    int shardId;
+    if (shardSuggestion != null && shardSuggestion < channels.length) {
+      shardId = shardSuggestion;
```
`ShardSuggestion` being null is a much different condition than it exceeding length of `channels`.
In the second case I'd expect at least a warning.

Bouncheck replied:
I've wondered if i can skip this check entirely, because it should not really happen unless someone changes node configuration so that specifically this does happen, but I've added the warning.

Lorak-mmk reviewed (changes requested):
I left some comments. Some things that fit better here:
- Why are there no tests for tablets? This is a complex feature so it should have proper tests (both unit tests of the new data structure and integration tests that check that tablet routing works properly in various scenarios)
- I see you decided to store replicas `Uuid`s instead of `Node` objects. This has a performance cost, because you need to allocate a new `HashSet` of all reploca `Node` objects for each query. I assume it was done to simplify the code, because you don't need to handle the case where we can't map some id to `Node`, like I do in maintenance procedures in Rust Driver Tablets PR (https://github.com/scylladb/scylla-rust-driver/pull/937). But the maintenance is still required for cases like node being replaces (lack of handling of this case is another problem with this PR), so handling the missing-UUID case doesn't really add much complexity.
- The PR is one being commit. It should be split into small atomic commits to make it easier to analyze fopr reviewers and future readers.

bhalevy commented:
ping
I see review comments from @Lorak-mmk from ~3 weeks ago that weren't discussed.

Bouncheck commented:
Yes, I've seen them and I'm reworking those parts. Should be ready soon

Bouncheck committed:
Tablets support

Introduces basic tablets support for version 4.x of the driver.
Metadata about tablets will be kept in TabletMap that gets continuously updated
through the tablets-routing-v1 extension. Each time the BoundStatement targets
the wrong node and shard combination the server supporting tablets should
respond with tablet metadata inside custom payload of its response.
This metadata will be transparently processed and used for future queries.

Tablets metadata will by enabled by default. Until now driver was using
TokenMaps to choose replicas and appropriate shards. Having a token was enough
information to do that. Now driver will first attempt tablet-based lookup
and only after failing to find corresponding tablet it will defer to TokenMap
lookup. Since to find a correct tablet besides the token we need the keyspace
and table names, many of the methods were extended to also accept those
as parameters.

RequestHandlerTestHarness was adjusted to mock also MetadataManager.
Before it used to mock only `session.getMetadata()` call but the same can
be obtained by `context.getMetadataManager().getMetadata()`. Using the
second method was causing test failures.

Bouncheck commented:
I'll push the code likely today - I want to run 1 more concurrency experiment.

roydahan commented:
@Bouncheck is it ready for re-review?
If so, please request @avelanarius and @Lorak-mmk reviewing it ASAP.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,288 @@
+        Set<Node> replicaNodes,
+        Map<Node, Integer> replicaShards) {
+      this.keyspaceName = keyspaceName;
+      this.tableId = tableId;
```
What do we need it for? I don't see any usages.

Bouncheck replied:
At one point I was told to leave this field as it is, but I don't remember what were the plans what to use it for in the future.
I guess I can remove it and worst case reimplement later.

Bouncheck replied:
Removed

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,288 @@
+          firstToken,
+          lastToken);
+
+      Set<NodeShardPair> replicas = new HashSet<>();
+      Set<Node> replicaNodes = new HashSet<>();
+      Map<Node, Integer> replicaShards = new HashMap<>();
+      List<TupleValue> list = tupleValue.getList(2, TupleValue.class);
```
Why do we need 3 collections? `replicaNodes` is required for easier usage in LBP (otherwise we'd need to create a new set on each query) but `replicas` and `replicaShards` seem redundant.

Bouncheck replied:
`replicaShards` was for easier, faster lookup of shards and `replicas` is now redundant.

Bouncheck replied:
Removed `replicas`. `replicaShards` stayed as necessary for shard lookup.

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,288 @@
+
+    @VisibleForTesting
+    DefaultTablet(
+        String keyspaceName,
+        UUID tableId,
+        String tableName,
```
Why do we need this for each tablet? It is going to be the same for each tablet in a table, so it this duplication seems like a memory waste

Bouncheck replied:
Most of the fields are there because I've been just copying the structure I've seen in design doc.
The only function this has right now is that it simplifies passing it around. I can remove this, but I'll have to pass keyspace and table as separate arguments. This will still save memory because it won't be preserved long term inside the map.

Bouncheck replied:
Removed those fields

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/KeyspaceTableNamePair.java:
```diff
@@ -0,0 +1,47 @@
+public class KeyspaceTableNamePair {
+  private final String keyspace;
+  private final String tableName;
+
```
Can those be null? If not, shouldn't there be `@NonNull` attributes?

Bouncheck replied:
Added NonNull annotations

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/api/core/metadata/NodeShardPair.java:
```diff
@@ -0,0 +1,28 @@
+ * guarantees about existence of a shard on corresponding host.
+ */
+public class NodeShardPair {
+  private final Node node;
```
ditto

Bouncheck replied:
I removed this class as it was no longer needed

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,288 @@
+  @Override
+  public void addTablet(Tablet tablet) {
+    KeyspaceTableNamePair ktPair =
+        new KeyspaceTableNamePair(tablet.getKeyspaceName(), tablet.getTableName());
+
+    // Get existing tablets for given table
+    NavigableSet<Tablet> existingTablets =
+        mapping.computeIfAbsent(ktPair, k -> new ConcurrentSkipListSet<>());
+    // Single tablet token range is represented by (firstToken, lastToken] interval
+    // We need to do two sweeps: remove overlapping tablets by looking at lastToken of existing
+    // tablets
+    // and then by looking at firstToken of existing tablets. Currently, the tablets are sorted
+    // according
+    // to their lastTokens.
+
+    // First sweep: remove all tablets whose lastToken is inside this interval
+    Iterator<Tablet> it = existingTablets.headSet(tablet, true).descendingIterator();
+    while (it.hasNext()) {
+      Tablet nextTablet = it.next();
+      if (nextTablet.getLastToken() <= tablet.getFirstToken()) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Second sweep: remove all tablets whose firstToken is inside this tuple's (firstToken,
+    // lastToken]
+    // After the first sweep, this theoretically should remove at most 1 tablet
+    it = existingTablets.tailSet(tablet, true).iterator();
+    while (it.hasNext()) {
+      Tablet nextTablet = it.next();
+      if (nextTablet.getFirstToken() >= tablet.getLastToken()) {
+        break;
+      }
+      it.remove();
+    }
+
+    // Add new (now) non-overlapping tablet
+    existingTablets.add(tablet);
+  }
```
This method is still susceptible to race conditions: 2 overlapping tablets can be added.
First both calls go trough sweeps
Then both go to `existingTablets.add(tablet);`, and add overlapping tablets.

Such races are one of the main reasons why I went with a totally different design in Rust Driver where there are no concurrent accesses to the structure, so no locks needed and no races possible.

I don't immediately see how to improve this data structure to fix races without introducing lock for the whole duration of `addTablet` call. Adding such a lock would imo not match assumptions of this driver - see `manual/developer/common/concurrency/README.md`, especially `Non-blocking` section and `manual/core/non_blocking/README.md`.

I don't see any solutionother than making the design similar to Rust Driver, unless you can somehow make this structure thread safe without locks.
If we were to change the design:
Does java driver have any internal threads that could be used to implement solution similar to Rust Driver? If not, can we create such a thread?

Bouncheck replied:
The structure itself is susceptible, but the way driver uses it is as described in 'Cold path' section of concurrency readme you mention. Any modifications go through `adminExecutor`s thread.

Lorak-mmk replied:
Then why does it need concurrent hash maps? They are much slower than normal ones and unnecessary if there are no concurrent accesses

Bouncheck replied:
There are also multiple threads just reading the structure and according to javadoc "If multiple threads access a map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally. (A structural modification is any operation that adds or deletes one or more mappings; merely changing the value associated with an existing key is not a structural modification.)"

Bouncheck replied:
That's quoted from the page for the regular hashmap of course. Situation is the same with sets.

Lorak-mmk replied:
Ok, in that case please explain in the comment:
- Thread (un-)safety of this structure
- Why this (un-)safety it's not violated: who calls it etc

Lorak-mmk replied:
One more question: 

> The structure itself is susceptible, but the way driver uses it is as described in 'Cold path' section of concurrency readme you mention. Any modifications go through `adminExecutor`s thread.

Is this a new thing? I remember that previous version of this PR was copying tablet list when addind new tablet and comment said that it is to prevent concurrent modification

Bouncheck replied:
Sorry, I did not see your followup question before. It is in a sense that it finally works now. It was in a half prepared state for that before (in v1, assuming right now is v3) and I did not notice that there were non-admin threads handling tablet inserts.

Bouncheck replied:
Added comment about thread safety to DefaultTabletMap class

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTabletMap.java:
```diff
@@ -0,0 +1,288 @@
+        break;
+      }
+      it.remove();
+    }
```
The code for removal seems correct, but the comments are not.
According to comments we would not remove (10, 70] tablet when inserting (20, 50] tablet , because neither 10 nor 70 is in (20, 50] range.

Bouncheck replied:
Fixed the comment. Double checked that the method does what it was supposed to.

Lorak-mmk commented on integration-tests/src/test/java/com/datastax/oss/driver/core/metadata/DefaultMetadataTabletMapIT.java:
```diff
@@ -0,0 +1,158 @@
+    // preparedStatement.enableTracing();
+    int counter = 0;
+    for (int i = 1; i <= QUERIES; i++) {
+      ResultSet rs = session.execute(preparedStatement.bind(i, i).setTracing(true));
```
Where is the tracing used in this test?

Bouncheck replied:
It is not. Removed tracing

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/api/core/cql/PreparedStatement.java:
```diff
@@ -58,6 +59,13 @@ public interface PreparedStatement {
   @NonNull
   ColumnDefinitions getVariableDefinitions();
 
+  /**
+   * Table name inferred from {@link PreparedStatement#getVariableDefinitions()} or {@code null} if
+   * not possible.
+   */
+  @Nullable
+  CqlIdentifier inferTable();
+
```
The fact that this is defined in different place than keyspace still feels extremely weird to me.
With tablets the table is a part of routing info, exactly the same as keyspace.
I don't see why they would be handled differently- is there some difference that justifies it? Maybe keyspace is not only used for routing but for something different too?


Apart from placement of this method there are other difference: the driver has a concept of routing keyspace,  right? Why isn't routing table introduced then?

Please explain why the differences are present.

Bouncheck replied:
Main motivation was keeping new additions as specific as possible and not extending what I don't need to. `PreparedStatement` has 1 implementation while `Request` has 20 implementations. At first I thought it will be easier to maintain this way, but I may as well provide default implementation that returns null and overrides in specific implementations. This way it should not complicate writing implementations of Request. 

Routing table can be introduced. I probably should add field and setters too in that case.

Bouncheck replied:
Moved this method to `Request` as `getRoutingTable` and provided override for `BatchStatement` and `BoundStatement`

Lorak-mmk commented on core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java:
```diff
@@ -321,6 +332,27 @@ protected Set<Node> getReplicas(@Nullable Request request, @Nullable Session ses
       return Collections.emptySet();
     }
 
+    if (table != null && tabletMap != null) {
```
I still don't get when tabletMap can be null

Bouncheck replied:
I've missed this check during previous iteration. Removed.

mykaul commented:
@dani-tweig - this important PR is not tracked anywhere - there's no 'tablets' label for it, etc.

Bouncheck commented:
Pushed another version:
- Removed redundant/unused fields and collections from DefaultTablet
- Added NonNull annotation to pair structures
- Removed NodeShardPair altogether
- Added comment about thread safety to DefaultTabletMap
- Fixed comments about sweep during tabletAdd in DefaultTabletMap
- Removed unused elements from tests
- Removed `inferTable`, added `getRoutingTable` in `Request`; Provided overrides for `BatchStatement` and `BoundStatement`. Ultimately I did not add `setRoutingTable` or `routingTable` fields. I do not have a usecase for them right now and this saves me from adjusting statement builders and tests.
- Removed unnecessary nullity checks and other misc. items pointed out
- Reduced 1 partitioner.hash(key) call in BasicLoadBalancingPolicy in case when falling back to TokenMap 
- Switched to CqlIdentifiers instead of Strings when handling keyspace and table in tablet contexts

Lorak-mmk commented:
I see CI is failing, do you know why? Also, I remember you mentioned that Tablet tests were not being executed, because of some edge case with version comparing. Is hat fixed?

Bouncheck commented:
Last time I checked this test these same two methods (`should_apply_node_filter`, `should_use_round_robin_on_local_dc_when_not_enough_routing_information`) were failing and it was due to one node failing to start. This changes actual number of nodes seen in local dc and changes the output in second method to different than expected because round robin does not include that downed node.

The problem with tablets integration test was that the CI used to pull release candidate version for which I've noticed it was skipping this test. It's pulling 6.0.0 now which is the minimal required version and if you search the logs for `DefaultMetadataTabletMapIT` you can see that it indeed does run the test and launch 2 nodes it's configured for.
